src/RedisIPCMessenger.ts

```diff
@@ -0,0 +1,148 @@
+import Redis from 'ioredis';
+import IPCMessenger, {
+  Room,
+  Instance,
+  MessageCallback,
+  MessageTypes,
+  Message,
+} from './IPCMessenger';
+
+type ConstructorParams = {
+  instance: Instance;
+  redis?: Redis.RedisOptions,
+  expireTime?: number,
+  refreshInterval?: number,
+}
+
+class RedisIPCMessenger implements IPCMessenger {
+  private readonly instance: Instance;
+  private readonly publisher: Redis.Redis;
+  private readonly subscriber: Redis.Redis;
+  private readonly expireTime: number;
+  private readonly refreshInterval: number;
+  private subscriptions: Map<Room, MessageCallback>;
+  private hasSetupCallbacks: boolean;
+
+  constructor(params: ConstructorParams) {
+    this.validateInstance(params.instance);
+
+    this.instance = params.instance;
+    this.subscriptions = new Map();
+    this.publisher = new Redis(params.redis);
+    this.subscriber = new Redis(params.redis);
+    this.expireTime = params.expireTime ?? 30000;
+    this.refreshInterval = params.refreshInterval ?? 10000;
+    this.hasSetupCallbacks = false;
+  }
+
+  async join(room: Room, callback: MessageCallback): Promise<void> {
+    this.validateRoom(room);
+
+    this.storeSubscription(room, callback);
+    await this.setupCallbacksIfNecessary();
+    await this.subscribeForRoomEvents(room);
+    await this.startRefreshKeyLoop(room);
+  }
+
+  async getOtherInstances(room: Room): Promise<Array<Instance>> {
+    this.validateRoom(room);
+    this.makeSureRoomIsJoined(room);
+
+    const keys = await this.publisher.keys(`${room}:*`);
+    return keys.map((key) => {
+      const [, instance] = key.split(':');
+      return instance;
+    }).filter((instance) => instance !== this.instance);
+  }
+
+  async send(room: Room, message: Omit<Message, 'sender'>): Promise<void> {
+    this.validateRoom(room);
+    this.makeSureRoomIsJoined(room);
+
+    await this.publisher.publish(room, this.serialize({
+      ...message,
+      sender: this.instance,
+    }));
+  }
+
+  private validateRoom(room: string) {
```

Comment:
This doesn't make much sense - there is a `Room` type, which is an alias for `string`, the constructor is using the `Room` type, but it's then passed to this method here, which implicitly treats it as a `string`. 

The approach should be fundamentally different - you have a type, so you should create a type guard on the type level and use that in the class instead of a private validator.

Reply:
Well, thinking about it, you don't need any validation within the class itself - as long as you export the typeguard, your users will be required to do the type coercion in advance.

Reply:
Something like this (in the types file):

```js
export const makeRoom  = (name: string): Room => {
  if (!name.length || name.includes(':')) {
    throw new TypeError(`A room name must be a non empty string and must not include ':', ${name} received`);
  }

  return name as Room;
};

export const isRoom = (r: string): r is Room => (!!r.length && r.includes(':'));```

Reply:
It's a leftover from when the method was called `validateRoomOrInstance`

Reply:
The rest of the comments are valid - you're not using TS as much as you can.

Reply:
Did something similar, omitted the exporting the validators though.

Reply:
The above can further be improved to avoid hard casts completely:

```ts
export const makeRoom  = (name: string): Room => {
  if (isRoom(name)) { return name; }
  throw new TypeError(`A room name must be a non empty string and must not include ':', ${name} received`);
};

export const isRoom = (r: string): r is Room => (!!r.length && r.includes(':'));
```

Reply:
Do you suggest using `isRoom` anywhere inside the client? Because atm there is no need for it.

Reply:
No - why would we do that? We would expect that `Room` is actually of the correct type. If it isn't, it's the caller's problem - they shouldn't have used hard casting, in favor of `makeRoom`.

Reply:
In this case, if the `isRoom` is used only to omit the `as`, I don't see much point in having it.

Reply:
Ended up having `isWhatever`, as it makes the throwing much shorter.